File: benchmark.cpp
#include <iostream>
#include <fstream>
#include <vector>
#include <string>
#include <chrono>
#include "sorting.h"

std::vector<int> readFile(const std::string& path) {
    std::ifstream f(path);
    int n; f >> n;
    std::vector<int> data(n);
    for (int& v : data) f >> v;
    return data;
}

struct Result {
    std::string dataset;
    std::string algorithm;
    int n;
    long long comparisons;
    long long moves;
    double ms;
};

void runOne(const std::string& dataset, const std::string& algName,
            void (*fn)(std::vector<int>&, int, int, Metrics&),
            const std::vector<int>& original, std::vector<Result>& results) {
    std::vector<int> arr = original;
    Metrics m;
    auto start = std::chrono::high_resolution_clock::now();
    fn(arr, 0, (int)arr.size() - 1, m);
    auto end = std::chrono::high_resolution_clock::now();
    double ms = std::chrono::duration<double, std::milli>(end - start).count();

    if (!std::is_sorted(arr.begin(), arr.end())) {
        std::cerr << "ERROR: " << algName << " en " << dataset << " no ordeno bien\n";
    }
    results.push_back({dataset, algName, (int)arr.size(), m.comparisons, m.moves, ms});
}

int main() {
    std::vector<std::string> datasets = {"data_random.txt", "data_sorted.txt",
                                          "data_reverse.txt", "data_partial.txt"};
    std::vector<Result> results;

    for (const auto& ds : datasets) {
        std::vector<int> data = readFile(ds);
        runOne(ds, "MergeSort2", mergeSort2, data, results);
        runOne(ds, "MergeSort3", mergeSort3, data, results);
        runOne(ds, "QuickSort1", quickSort1, data, results);
        runOne(ds, "QuickSort2", quickSort2, data, results);
    }

    std::ofstream csv("resultados.csv");
    csv << "dataset,algoritmo,n,comparaciones,intercambios,tiempo_ms\n";
    for (const auto& r : results) {
        csv << r.dataset << "," << r.algorithm << "," << r.n << ","
            << r.comparisons << "," << r.moves << "," << r.ms << "\n";
    }
    csv.close();

    printf("%-16s %-11s %8s %14s %14s %10s\n",
           "Dataset", "Algoritmo", "N", "Comparaciones", "Intercambios", "Tiempo(ms)");
    for (const auto& r : results) {
        printf("%-16s %-11s %8d %14lld %14lld %10.3f\n",
               r.dataset.c_str(), r.algorithm.c_str(), r.n, r.comparisons, r.moves, r.ms);
    }

    std::cout << "\nResultados guardados en resultados.csv\n";
    return 0;
}

File: benchmark.py
import sys
import time
import csv
import threading

from sorting import Metrics, merge_sort2, merge_sort3, quick_sort1, quick_sort2

sys.setrecursionlimit(2_000_000)


def read_file(path):
    with open(path) as f:
        f.readline()  # n, no se usa directo
        data = list(map(int, f.readline().split()))
    return data


def run_one(dataset, name, func, original, results):
    arr = original.copy()
    m = Metrics()

    def target():
        func(arr, 0, len(arr) - 1, m)

    start = time.perf_counter()
    t = threading.Thread(target=target)
    t.start()
    t.join()
    elapsed_ms = (time.perf_counter() - start) * 1000

    if any(arr[i] > arr[i + 1] for i in range(len(arr) - 1)):
        print(f"ERROR: {name} en {dataset} no ordeno bien")

    results.append((dataset, name, len(arr), m.comparisons, m.moves, elapsed_ms))


def main():
    threading.stack_size(64 * 1024 * 1024)  # stack grande para recursion profunda (peor caso)

    datasets = ["data_random.txt", "data_sorted.txt", "data_reverse.txt", "data_partial.txt"]
    results = []

    for ds in datasets:
        data = read_file(ds)
        run_one(ds, "MergeSort2", merge_sort2, data, results)
        run_one(ds, "MergeSort3", merge_sort3, data, results)
        run_one(ds, "QuickSort1", quick_sort1, data, results)
        run_one(ds, "QuickSort2", quick_sort2, data, results)

    with open("resultados_python.csv", "w", newline="") as f:
        w = csv.writer(f)
        w.writerow(["dataset", "algoritmo", "n", "comparaciones", "intercambios", "tiempo_ms"])
        w.writerows(results)

    print(f"{'Dataset':<16} {'Algoritmo':<11} {'N':>8} {'Comparaciones':>14} {'Intercambios':>14} {'Tiempo(ms)':>10}")
    for r in results:
        print(f"{r[0]:<16} {r[1]:<11} {r[2]:>8} {r[3]:>14} {r[4]:>14} {r[5]:>10.3f}")

    print("\nResultados guardados en resultados_python.csv")


if __name__ == "__main__":
    main()